### ISSUE TITLE
Random oscilation or not?

### ISSUE THREAD

**Issue by agrignard** [CityScope/CS_Cooper-Hewitt#9]
@yasushisakai briefly regarding this commit https://github.com/CityScope/CS_Cooper-Hewitt/commit/3d64c4430b7e4dead1a045d66dce94ce5be2d8b9

I think this modif

replacing 
````
            if ( path.indexOf(toNode) == 0 ) {  
              pos = destNodePos;
              srcNode =  (Node) map.graph.nodes.get(int(random(map.graph.nodes.size())));
              destNode =  (Node) map.graph.nodes.get(int(random(map.graph.nodes.size())));
              pos= new PVector(srcNode.x, srcNode.y);
              path=null;
              dir = new PVector(0.0, 0.0);
`````
by
`````
              // Not destination. Look for next node --->
            if (path.indexOf(toNode) == 0 ) {  
              pos = destNodePos; // ?
              this.initAgent();
``````

made the agent not oscilatting anymore. I am going to fix this I just wanted to know if it is by purpose or not? at the beginning I think some random oscilatting agent are helpful (instaed of having them stopping.

**Comment by agrignard** [comment]
maybe also due to this change

`````
    srcNode =  (Node) map.graph.nodes.get(int(random(map.graph.nodes.size())));
    destNode =  (Node) map.graph.nodes.get(int(random(map.graph.nodes.size())));
    pos= new PVector(srcNode.x,srcNode.y);
    path=null;
``````
by 
`````
     boolean isAdjecent = true;
     while(srcNode == destNode || isAdjecent) {
      srcNode =  (Node) map.graph.nodes.get(int(random(map.graph.nodes.size())));
      destNode =  (Node) map.graph.nodes.get(int(random(map.graph.nodes.size())));
      isAdjecent = srcNode.connectedTo(destNode);
    }    
    
    pos = new PVector(srcNode.x,srcNode.y);
    path = null;
``````

**Comment by yasushisakai** [comment]
this was partially to detect sanity of the geojson file. 

1. Your first comment, what we were doing and  ```initAgent()``` was the same, so the decision to reduce redundancy. (BTW, why are we having a function name init, when we can put it in the constructor?)

2. The second has two details. 
- ```isAdjecent``` comes from me debugging the geojson file, our network has mono directional paths, so some nodes has no way to come back (which was resolved by changing the structure of the network). Thus irrelevant now.

- The ```while``` loop was to guarantee that it's not pointing it self as src and dest, but I noticed that consequently ```initAgent``` will not update its Nodes after the first call, because the src and dest will always be different. -> I changed to a do while loop

**Comment by agrignard** [comment]
This commit was suppose to put the infinite walk
https://github.com/CityScope/CS_Cooper-Hewitt/commit/fbebafedbabc9aeb8ed2d36674e80573753cc88b

**Comment by agrignard** [comment]
ok so one of the difference is that in the current case we don't go in this part of the code (compare to the working version)

  private boolean calcRoute(Node origin, Node dest) {
    // Agent already in destination --->
    if (origin == dest) {
      toNode=dest;
      return true;
      // Next node is available --->
    }

**Comment by agrignard** [comment]
ok I find the guilty

public void run(PGraphics p){
    for (Agent agent : agents) {
      agent.move(1);
      agent.draw(p);
    }
  }

**Comment by agrignard** [comment]
this one is pretty hard to track and it seems (but maybe not) to be link to the intricution of the mutli threading.  I am almost thinking to restart from a working version https://github.com/CityScope/CS_Cooper-Hewitt/commit/fbebafedbabc9aeb8ed2d36674e80573753cc88b  and then iterate on this in order to be sure that any new commit doesn't braek the current behavior. Even if this is pretty far in term of commit, as the path has already be made maybe it's an option to remake it in order to iterate on a clean version.

**Comment by yasushisakai** [comment]
that's fine, lets. do it!

The commit before 3d64c4430b7e4dead1a045d66dce94ce5be2d8b9 has a NPE error in line 111.

``` toNode = path.get(path.size()-2);```

The class structure needs to be refined anyway. (let's talk this later)

**Comment by agrignard** [comment]
... so hard to track...

Is it possible that those change has an impact?

https://github.com/CityScope/CS_Cooper-Hewitt/commit/3d64c4430b7e4dead1a045d66dce94ce5be2d8b9#diff-1ffd559a41afca715a7fb2e86f66e937

**Comment by agrignard** [comment]
I potentially found it!!!!!!!!

What is the reason for this very sioux modif ?????? 
replacing this 
`````
        PVector pos = toXY(points.getJSONArray(j).getFloat(1), points.getJSONArray(j).getFloat(0));
`````
by 
`````
        PVector pos = toXY(points.getJSONArray(j).getFloat(0), points.getJSONArray(j).getFloat(1));
`````


        PVector pos = toXY(points.getJSONArray(j).getFloat(1), points.getJSONArray(j).getFloat(0));
        PVector pos = toXY(points.getJSONArray(j).getFloat(0), points.getJSONArray(j).getFloat(1));

by replacing the original code the agent doesn't stop anymore but the graph seems offsetted or misplaced. 

It's getting there

**Comment by yasushisakai** [comment]
I take this.

**Comment by yasushisakai** [comment]
resolved at commit: ded9e69ce04f4401b214266e697eeb7d3bf2ebc6
